quantum/action_tapping.c

```diff
@@ -708,6 +734,110 @@ void waiting_buffer_scan_tap(void) {
     }
 }
 
+#    ifdef SPECULATIVE_HOLD
+__attribute__((weak)) bool get_speculative_hold(uint16_t keycode, keyrecord_t *record) {
+    return (QK_MOD_TAP_GET_MODS(keycode) & (MOD_LALT | MOD_LGUI)) == 0;
+}
+
+static void debug_speculative_keys(void) {
+    ac_dprintf("mods = { ");
+    for (int8_t i = 0; i < num_speculative_keys; ++i) {
+        ac_dprintf("%02X ", speculative_keys[i].mods);
+    }
+    ac_dprintf("}, keys = { ");
+    for (int8_t i = 0; i < num_speculative_keys; ++i) {
+        ac_dprintf("%02X%02X ", speculative_keys[i].key.row, speculative_keys[i].key.col);
+    }
+    ac_dprintf("}\n");
+}
+
+static void process_speculative_press(keyrecord_t *record) {
+    if (num_speculative_keys >= SPECULATIVE_KEYS_SIZE) { // Overflow!
+        ac_dprintf("SPECULATIVE KEYS OVERFLOW: IGNORING EVENT\n");
+        return;
+    }
+
+    const uint16_t keycode = get_record_keycode(record, true);
+    if (!IS_QK_MOD_TAP(keycode)) {
+        return;
+    }
+
+    uint8_t mods = QK_MOD_TAP_GET_MODS(keycode);
+    if ((mods & 0x10) != 0) { // Unpack 5-bit mods to 8-bit representation.
+        mods <<= 4;
+    }
+
+    if ((~(get_mods() | get_speculative_mods()) & mods) != 0 && get_speculative_hold(keycode, record)) {
+        add_speculative_mods(mods);
+        speculative_keys[num_speculative_keys] = (speculative_key_t){
+            .key  = record->event.key,
+            .mods = mods,
+        };
+        ++num_speculative_keys;
+
+        ac_dprintf("Speculative Hold: ");
+        debug_speculative_keys();
+    }
+}
+
+void process_speculative_hold(keyrecord_t *record) {
```

Comment:
Please add a comment noting that this is called at the top of process_record and in action_tapping.c to be alongside the other logic. I would also rename it to be more descriptive, personally - you are matching the naming of process_record_quantum functions, but flow_tap_update_last_event is already bailing on that for process_record and this is really confusing with there being a separate process_speculative_press in this file.